### ISSUE TITLE
Partially unknown types warning issued for incomplete types related to runtime optional dependencies

### ISSUE THREAD

**Issue by DervishD** [microsoft/pyright#11069]
## Environment data
-   Pylance version: 2025.9.1
-   OS and version: Windows 11 Home 25H2
-   Python version (& distribution if applicable, e.g. Anaconda): 3.14.0

## Code Snippet
```python
from PIL.Image import open as open_image, Resampling

test_image = open_image('test.jpg') 
resized_image = test_image.resize((800, 600), Resampling.LANCZOS)  # Warning on this line.
resized_image.save('resized_test.jpg')
test_image.close()
```

## Expected behavior

No warning issued for the marked line.

## Actual behavior
This warning is issued for the marked line.
```
Type of "resize" is partially unknown
  Type of "resize" is "(size: tuple[int, int] | list[int] | Unknown, resample: int | None = None, box: tuple[float, float, float, float] | None = None, reducing_gap: float | None = None) -> Image"
```

The reason is that `resize()` has this signature:
```python
def resize(
        self,
        size: tuple[int, int] | list[int] | NumpyArray,
        resample: int | None = None,
        box: tuple[float, float, float, float] | None = None,
        reducing_gap: float | None = None,
    ) -> Image:
```
and in turn, `NumpyArray` is defined as this:
```python
if TYPE_CHECKING:
    from numbers import _IntegralLike as IntegralLike

    try:
        import numpy.typing as npt

        NumpyArray = npt.NDArray[Any]  # requires numpy>=1.21
    except (ImportError, AttributeError):
        pass
```

That is, an optional type-checking dependency is causing a function to have an incomplete signature, by introducing an `Unknown` type. I don't know whether is debatable that Pylance should ignore such situations, removing the optionally `Unknown` type, or if the library should implement some kind of workaround.

I have not been able to propose a solution because buildling an optional type in case of `ImportError` does not seem to work for some reason (the type created on the `try:` branch seems to take precedence over the one defined in the `except:` branch, just like the `import` has not failed):

```python
if TYPE_CHECKING:
    from numbers import _IntegralLike as IntegralLike

    try:
        import numpy.typing as npt

        NumpyArray = npt.NDArray[Any]  # requires numpy>=1.21
        SizeType = tuple[int, int] | list[int] | NumpyArray  # Is the one used by Pylance even when ImportError is raised.
    except (ImportError, AttributeError):
        SizeType = tuple[int, int] | list[int]
```

In short, I don't know if this should be fixed or not, and where should it be fixed if is a bug/improvement.

As a reference, my original report in the Pillow repository:  https://github.com/python-pillow/Pillow/issues/9280

## Logs
I think logs are irrelevant here, because the reason of the warning is known.

**Comment by erictraut** [comment]
The `reportUnknownXXX` checks are not enabled by default in pyright. They are strict-mode checks. I presume that you've manually enabled them in your project.

Pyright is working as intended in this case. It's telling you that the type of `resize` is partially unknown. You've diagnosed the underlying reason — one of the symbols used in the signature is not defined. If you want to eliminate the "unknown" diagnostic, you'll need to address this.

**Comment by DervishD** [comment]
Thanks for the reply, @erictraut and yes, I've enabled strict mode. And I cannot eliminate the diagnostic because the problem is being caused by a third-party library, so my best bet is to disable analysis for the offending line in _my_ code, which is not a big deal for me. Still, I'm modifying the third-party code to find a suitable solution.

Anyway the underlying reason does not appear to be fixable even for the authors of the third-party library, because apparently pyright is considering only the `try:` branch and not the `except:` branch of the solution I tried at type-checking time, so it looks like it's impossible to conditionally create a new type which does not include the one that would not be defined in the original `try/except` block.

IF I can prove this is the case, that the code inside the `try/except` block is not properly processed at type-checking time, would you accept the bug report? Or is this intended behavior? If that is the case I won't bother you anymore with this 😉

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

When `try/except` is used in code, a type checker cannot determine which path will be taken at runtime. Pyright assumes that the `try` branch will succeed. This heuristic is correct for most libraries. Changing it would result in a lot of breakage for typed code bases.

Library authors can clarify their intent for static type checkers by using code that's guarded by `if TYPE_CHECKING:`. This way, no heuristics or assumptions are needed on the part of the type checker.

**Comment by DervishD** [comment]
Thanks **A LOT** @erictraut for explaining that and for your help. I was suspecting that was the case and makes absolute sense. I'll try to find a solution to this, not because it needs a fix, but because I think it's very interesting to have a more or less general solution for the problem of having conditional types in `try/except`. Having incomplete types as a result looks like a hack to me, and the library should be fixed.

Thanks for your help and for taking the time of explaining to me the heuristics used 😊 I owe you one 😉

**Comment by radarhere** [comment]
@erictraut if you'll permit me, in your greater experience in working with type hints, do you think that
a) a library shouldn't explicitly reference optional dependencies in type hints, or
b) users wishing to check type hints in strict-mode should install optional dependencies as required by the type hints?

**Comment by erictraut** [comment]
Option b is typically preferable. There are cases where this isn't feasible — such as when the optional dependency is incompatible with versions of Python that are supported by the library.

**Comment by DervishD** [comment]
Even though I have mixed feelings here, I think the easier solution is option b. In fact is what I did at first, but then I thought that installing an optional dependency was a patch, not a real solution, for a potential problem. And I still think that using optional dependencies in typing is not a good idea, but that is something to be discussed on Pillow, not here.

I now think Pyright should not change the heuristics, there has to be a better solution for this.
